### PR TITLE
Enable engine factory to be pluggable

### PR DESCRIPTION
**Pull request by jasontedor**
This commit enables the engine factory to be pluggable based on index settings used when creating the index service for an index.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+   interface EngineFactoryProvider extends Function<IndexSettings, Optional<EngineFactory>> {
+
+   }
+
+   /**
+    * The engine factory provider for this engine plugin.
+    *
+    * @return the engine factory provider
+    */
+   EngineFactoryProvider getEngineFactoryProvider();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: do we really need the indirection here and can't we simply have a single factory method:

```Java
public EngineFactory getEngineFactory(IndexSettings settings);
```

I don't understand why we have to have another indirection?

**Review by s1monw — CHANGES_REQUESTED** [review]
left a question / suggestion

File: core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java (hunk 1 above)
Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: @s1monw I pushed 6e2f3ec381bcbe34ad97c7dd126a38589242706f.

File: core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * When an index is created this method is invoked for each engine plugin. Engine plugins can inspect the index settings to determine
+     * whether or not to provide an engine factory for the given index. A plugin that is not overriding the default engine should return
+     * {@link Optional#empty()}. If multiple plugins return an engine factory for a given index the index will not be created and an
+     * {@link IllegalStateException} will be thrown during index creation.
+     *
+     * @return an optional engine factory
+     */
+    Optional<EngineFactory> getMaybeEngineFactory(IndexSettings indexSettings);
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: Why the use of Optional? In other places we have just used `null` to indicate there is no engine implementation being added. Also, can I don't think we should use the `Maybe` style of naming.  Again, in the pull based plugin APIs we have not done this anywhere else, and I don't think it really adds anything.

**Review by martijnvg — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java (hunk 2 above)
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: +1 to null and rename to `getEngineFactory`

Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: I prefer the use of optional, but I'm fine with renaming.

Comment (rjernst) [reply core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: I think we should come to an agreement on whether we should use Optional for these, and be consistent. But arbitrarily adding inconsistencies into the API will just create confusion.

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by rjernst** [comment]
LGTM too. We can discuss Optional vs null next week.

File: core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java (hunk 2 above)
Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java]: I agree on the consistency argument. We discussed `Optional` versus `null` in a separate channel that we will merge this as-is for now (with `Optional`), consider changing the handful of other places in the plugin API where we allow `null` to be returned but will have a broader discussion about this first. We can return to this change if needed.

**Pull request merged**